torch/csrc/jit/passes/symbolic_shape_cache.cpp

```diff
@@ -0,0 +1,214 @@
+  auto ss_map = std::unordered_map<int64_t, int64_t>();
+  auto cache_key = get_cache_key(schema, arg_vec, ss_map, /* deep_copy */ true);
+  auto can_ret_vec = std::make_shared<std::vector<CanonicalizedSymbolicShape>>(
+      cannonicalizeVec(ret_vec, ss_map));
+  shapeCache.Add(cache_key, can_ret_vec);
+}
+
+TORCH_API c10::optional<std::vector<at::SymbolicShape>>
+get_cached_shape_function(
+    const FunctionSchema* schema,
+    const std::vector<SSAInput>& arg_vec) {
+  // TODO: compare perf using std::vector<std::tuple<int64_t, int64_t>> for both
+  // ss_map and inverse_ss_map
+  auto ss_map = std::unordered_map<int64_t, int64_t>();
+  auto cache_key =
+      get_cache_key(schema, arg_vec, ss_map, /* deep_copy */ false);
+  auto cached_ret_vec = shapeCache.Get(cache_key);
+  if (cached_ret_vec == nullptr) {
+    return c10::nullopt;
+  }
+  // Decanonicalize the return values
+  auto inverse_ss_map = std::unordered_map<int64_t, int64_t>();
+  for (auto& ss_val : ss_map) {
+    inverse_ss_map[ss_val.second] = ss_val.first;
+  }
+  std::vector<at::SymbolicShape> ret_vec;
+  for (auto& css : *cached_ret_vec) {
+    ret_vec.emplace_back(css.toSymbolicShape(inverse_ss_map));
+  }
+  return ret_vec;
+}
+
+// Function only to access the cache, used for testing
+TORCH_API void clear_shape_cache() {
+  shapeCache.Clear();
+}
+
+TORCH_API size_t get_shape_cache_size() {
+  return shapeCache.Numel();
+}
+
+void CanonicalizedSymbolicShape::init(
+    const c10::SymbolicShape& orig_shape,
+    std::unordered_map<int64_t, int64_t>& ss_map) {
+  auto sizes = orig_shape.sizes();
+  if (!sizes) {
+    values_ = c10::nullopt;
+    return;
+  }
+  values_ = std::vector<int64_t>();
+  int64_t cur_symbolic_index = -(int64_t)ss_map.size() - 1;
+  for (auto& cur_shape : *sizes) {
+    if (cur_shape.is_static()) {
+      values_->push_back(cur_shape.static_size());
+    } else {
+      // Check for aliasing
+      auto it = ss_map.find(cur_shape.value());
+
+      if (it == ss_map.end()) {
+        values_->push_back(cur_symbolic_index);
+        ss_map.insert({cur_shape.value(), cur_symbolic_index});
+        cur_symbolic_index--;
+      } else {
+        values_->push_back(it->second);
+      }
+    }
+  }
+}
+
+c10::SymbolicShape CanonicalizedSymbolicShape::toSymbolicShape(
+    std::unordered_map<int64_t, int64_t> inverse_ss_map) const {
+  if (!values_.has_value()) {
+    return c10::SymbolicShape();
+  }
+  std::vector<at::ShapeSymbol> sizes;
+  for (long long cur_val : *values_) {
+    if (cur_val >= 0) {
+      sizes.push_back(at::ShapeSymbol::fromStaticSize(cur_val));
+      continue;
+    }
+    auto res = inverse_ss_map.find(cur_val);
+    if (res != inverse_ss_map.end()) {
+      sizes.push_back(at::ShapeSymbol::fromStaticSize(res->second));
+    } else {
+      auto new_symbol = at::ShapeSymbol::newSymbol();
+      inverse_ss_map.insert({cur_val, new_symbol.value()});
+      sizes.push_back(new_symbol);
+    }
+  }
+  return c10::SymbolicShape(sizes);
```

Comment:
nit: `std::move(sizes)`. this is in hot path

Reply:
C++11 move semantics is quite complicated, correct me if I'm wrong, but adding `std::move()` seems to not do anything useful here unless we have both of these constructors defined:
```
SymbolicShape(std::vector<ShapeSymbol> dims); // Current
SymbolicShape(std::vector<ShapeSymbol>&& dims); // rvalue reference
```

I do however think that we likely want to replace the current constructor of SymbolicShape with the rvalue reference version.